cedar-policy-core/src/ast/name.rs

```diff
@@ -373,6 +386,110 @@ mod vars_test {
     }
 }
 
+/// A newtype which indicates that the contained [`Name`] does not contain
+/// reserved `__cedar`, as specified by RFC 52
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+pub struct UnreservedName(pub(crate) Name);
+
+impl Display for UnreservedName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl FromStr for UnreservedName {
+    type Err = ParseErrors;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let n: Name = s.parse()?;
+        n.try_into().map_err(ParseErrors::singleton)
+    }
+}
+
+impl FromNormalizedStr for UnreservedName {
+    fn describe_self() -> &'static str {
+        "Unreserved name"
+    }
+}
+
+impl UnreservedName {
+    /// Qualify the name with an optional namespace
+    /// This method has the same behavior as [`Name::qualify_with`] except that
+    /// the `namespace` argument is a `Option<&UnreservedName>`
+    pub fn qualify_with(&self, namespace: Option<&Self>) -> Self {
+        Self(self.as_ref().qualify_with(namespace.map(|n| n.as_ref())))
+    }
+
+    /// Create a [`UnreservedName`] with no path (no namespaces).
+    /// Returns an error if `s` is not a valid identifier.
+    pub fn parse_unqualified_name(s: &str) -> Result<Self, ParseErrors> {
+        Name::parse_unqualified_name(s).and_then(|n| n.try_into().map_err(ParseErrors::singleton))
+    }
+
+    /// Create a [`UnreservedName`] with no path (no namespaces).
+    pub fn unqualified_name(id: UnreservedId) -> Self {
+        Self(Name::unqualified_name(id.0))
+    }
+
+    /// Get the basename of the [`UnreservedName`] (ie, with namespaces stripped).
+    /// Return a reference to [`Id`]
+    pub fn basename_unchecked(&self) -> &Id {
+        self.0.basename()
+    }
```

Comment:
I'm not sure I love the `_unchecked()` suffix here, as usually we use that when the caller is responsible for upholding some invariant.  Here, there is no invariant to be upheld; we return `Id` which doesn't promise any particular invariant.  To me, it might make more sense to call this method `.basename()` and call the one below something else.

Reply:
I changed it to `basename_as_ref`.